### PR TITLE
update SSL Labs scanner with new API, be robust

### PR DESCRIPTION
**Pull request by busterb**
This updates the SSL Labs scanner to know about new additions to the API, and prevents the module from breaking again just because there is new JSON in the output. I couldn't figure out how to get the Api class to print messages normally, and there is some other output that needs to be added. But the module does work again.

## Verification

- [x] `./msfconsole -qx 'use gather/ssllabs_scan; set HOSTNAME google.com; run'`
- [x] **Verify** the module actually runs and prints results
- [x] **Verify** the module doesn't immediately exit with an Invalid Parameter error

### CODE REVIEW

File: modules/auxiliary/gather/ssllabs_scan.rb [hunk 1]
```diff
@@ -142,18 +156,30 @@ def self.has_object_ref(name, klass)
...
           obj.instance_variable_set("@#{name}", value)
         elsif @lists.key?(name)
-          obj.instance_variable_set("@#{name}", value.map { |v| @lists[name].load(v) }) unless value.nil?
+          unless value.nil?
+            var = value.map do |v|
+              val, ua = @lists[name].load(v)
+              unused_attrs.concat ua
+              val
+            end
+            obj.instance_variable_set("@#{name}", var)
```
Comment (sempervictus) [inline modules/auxiliary/gather/ssllabs_scan.rb]: How safe is this on externally supplied data?

File: modules/auxiliary/gather/ssllabs_scan.rb [hunk 2]
```diff
@@ -142,18 +156,30 @@ def self.has_object_ref(name, klass)
...
+          end
         elsif @refs.key?(name)
-          obj.instance_variable_set("@#{name}", @refs[name].load(value)) unless value.nil?
+          unless value.nil?
+            val, ua = @refs[name].load(value)
+            unused_attrs.concat ua
+            obj.instance_variable_set("@#{name}", val)
```
Comment (sempervictus) [inline modules/auxiliary/gather/ssllabs_scan.rb]: ^^

File: modules/auxiliary/gather/ssllabs_scan.rb (hunk 1 above)
Comment (busterb) [reply modules/auxiliary/gather/ssllabs_scan.rb]: Note that this is just an unwrapped version of the original one-liner. I don't see a risk given the amount of validation going on here. Are you thinking of a deserialization attack?

**Comment by busterb** [comment]
An alternative to fixing this module could be to delete it. It's pretty clear nobody had run it in a while, or had at least reported it being broken, as far as I can see.

**Comment by sempervictus** [comment]
Yeah, say someone owns ssllabs and msfconsole is run as root for priv binds... Makes a mess, so I generally shy away from :send, :eval, and the internal setters. I'm paranoid, this is not likely, but you never know who uses framework out there, and who wants to own them.

**Comment by busterb** [comment]
Interestingly, this module was broken precisely because it would refuse to continue if the input data didn't match its own model. This converts it to soldier on and simply discard extra data rather than fail outright. But caution is noted. This does seem to over-engineer the API access a tad, when data could maybe just be parsed on the fly.

**Comment by h00die** [comment]
I personally like the module, just didn't know it existed for a long while, and have been on a reporting binge for a few months

**Comment by busterb** [comment]
OK, I'm happy continuing to update it. I actually started poking at it because I was looking at a similar Komand plugin that couldn't invoke scans on its own to new hosts.

**Comment by h00die** [comment]
I can't find a way to break it.  I tried a bunch of sites, i'll save you the boring output other than metasploit.
Whenever I use the site for a client, I always do the 'Do not show the results on the boards'.  Interestingly, the API doesn't include a feature like that.  However, of the bunch of sites I scanned, none of them appeared on the site's list of recent sites.  Must be on by default or on a longer delay than I'm waiting for.
```
# ./msfconsole -qx 'use gather/ssllabs_scan; set HOSTNAME metasploit.com; set usecache false; run'
HOSTNAME => metasploit.com
usecache => false
[*] SSL Labs API info
[*] API version: 1.30.7
[*] Evaluation criteria: 2009o
[*] Running assessments: 0 (max 25)
[*] Server: metasploit.com - Resolving domain names
[*] Scanned host: 52.84.237.225 (server-52-84-237-225.sfo5.r.cloudfront.net)- 0% complete (Testing renegotiation)
[*] Ready: 0, In progress: 1, Pending: 7
[*] metasploit.com - Progress 0%
[*] Scanned host: 52.84.237.225 (server-52-84-237-225.sfo5.r.cloudfront.net)- 0% complete (Testing renegotiation)
...snip...
[*] Ready: 7, In progress: 1, Pending: 0
[*] metasploit.com - Progress 88%
[*] Scanned host: 52.84.237.237 (server-52-84-237-237.sfo5.r.cloudfront.net)- 92% complete (Testing Bleichenbacher)
[*] Ready: 7, In progress: 1, Pending: 0
[*] metasploit.com - Progress 88%
[*] Scanned host: 52.84.237.237 (server-52-84-237-237.sfo5.r.cloudfront.net)- 95% complete (Simulating handshakes)
[*] Ready: 7, In progress: 1, Pending: 0
[*] metasploit.com - Progress 88%
[*] Scanned host: 52.84.237.237 (server-52-84-237-237.sfo5.r.cloudfront.net)- 99% complete (Simulating handshakes)
[*] Ready: 7, In progress: 1, Pending: 0
[*] metasploit.com - Progress 88%
[*] Host: metasploit.com
[*] 	  52.84.237.225
[*] 	  52.84.237.145
[*] 	  52.84.237.234
[*] 	  52.84.237.21
[*] 	  52.84.237.113
[*] 	  52.84.237.53
[*] 	  52.84.237.48
[*] 	  52.84.237.237
[*] -----------------------------------------------------------------
[*] Report for server-52-84-237-225.sfo5.r.cloudfront.net (52.84.237.225)
[*] -----------------------------------------------------------------
[+] Overall rating: A+
[+] TLS 1.2 - Yes
[+] TLS 1.1 - Yes
[+] TLS 1.0 - Yes
[+] SSL 3.0 - No
[+] SSL 2.0 - No
[+] Secure renegotiation is supported
[!] BEAST attack - Yes
[+] POODLE SSLv3 - Not vulnerable
[+] POODLE TLS - Not vulnerable
[+] Downgrade attack prevention - Yes, TLS_FALLBACK_SCSV supported
[+] Freak - Not vulnerable
[+] RC4 - No
[*] Heartbeat (extension) - No
[+] Heartbleed (vulnerability) - No
[+] OpenSSL CCS vulnerability (CVE-2014-0224) - No
[+] Forward Secrecy - With modern browsers
[+] Strict Transport Security (HSTS) - Yes:max-age=15552000:includeSubdomains
[!] Public Key Pinning (HPKP) - No
[+] Compression - No
[*] Session resumption - No
[*] Session tickets - Yes
[*] OCSP Stapling - No
[*] Next Protocol Negotiation (NPN) - Yes (http/1.1)
[*] SNI Required - Yes
[*] -----------------------------------------------------------------
[*] Report for server-52-84-237-145.sfo5.r.cloudfront.net (52.84.237.145)
[*] -----------------------------------------------------------------
[+] Overall rating: A+
[+] TLS 1.2 - Yes
[+] TLS 1.1 - Yes
[+] TLS 1.0 - Yes
[+] SSL 3.0 - No
[+] SSL 2.0 - No
[+] Secure renegotiation is supported
[!] BEAST attack - Yes
[+] POODLE SSLv3 - Not vulnerable
[+] POODLE TLS - Not vulnerable
[+] Downgrade attack prevention - Yes, TLS_FALLBACK_SCSV supported
[+] Freak - Not vulnerable
[+] RC4 - No
[*] Heartbeat (extension) - No
[+] Heartbleed (vulnerability) - No
[+] OpenSSL CCS vulnerability (CVE-2014-0224) - No
[+] Forward Secrecy - With modern browsers
[+] Strict Transport Security (HSTS) - Yes:max-age=15552000:includeSubdomains
[!] Public Key Pinning (HPKP) - No
[+] Compression - No
[*] Session resumption - No
[*] Session tickets - Yes
[*] OCSP Stapling - No
[*] Next Protocol Negotiation (NPN) - Yes (http/1.1)
[*] SNI Required - Yes
[*] -----------------------------------------------------------------
[*] Report for server-52-84-237-234.sfo5.r.cloudfront.net (52.84.237.234)
[*] -----------------------------------------------------------------
[+] Overall rating: A+
[+] TLS 1.2 - Yes
[+] TLS 1.1 - Yes
[+] TLS 1.0 - Yes
[+] SSL 3.0 - No
[+] SSL 2.0 - No
[+] Secure renegotiation is supported
[!] BEAST attack - Yes
[+] POODLE SSLv3 - Not vulnerable
[+] POODLE TLS - Not vulnerable
[+] Downgrade attack prevention - Yes, TLS_FALLBACK_SCSV supported
[+] Freak - Not vulnerable
[+] RC4 - No
[*] Heartbeat (extension) - No
[+] Heartbleed (vulnerability) - No
[+] OpenSSL CCS vulnerability (CVE-2014-0224) - No
[+] Forward Secrecy - With modern browsers
[+] Strict Transport Security (HSTS) - Yes:max-age=15552000:includeSubdomains
[!] Public Key Pinning (HPKP) - No
[+] Compression - No
[*] Session resumption - No
[*] Session tickets - Yes
[*] OCSP Stapling - No
[*] Next Protocol Negotiation (NPN) - Yes (http/1.1)
[*] SNI Required - Yes
[*] -----------------------------------------------------------------
[*] Report for server-52-84-237-21.sfo5.r.cloudfront.net (52.84.237.21)
[*] -----------------------------------------------------------------
[+] Overall rating: A+
[+] TLS 1.2 - Yes
[+] TLS 1.1 - Yes
[+] TLS 1.0 - Yes
[+] SSL 3.0 - No
[+] SSL 2.0 - No
[+] Secure renegotiation is supported
[!] BEAST attack - Yes
[+] POODLE SSLv3 - Not vulnerable
[+] POODLE TLS - Not vulnerable
[+] Downgrade attack prevention - Yes, TLS_FALLBACK_SCSV supported
[+] Freak - Not vulnerable
[+] RC4 - No
[*] Heartbeat (extension) - No
[+] Heartbleed (vulnerability) - No
[+] OpenSSL CCS vulnerability (CVE-2014-0224) - No
[+] Forward Secrecy - With modern browsers
[+] Strict Transport Security (HSTS) - Yes:max-age=15552000:includeSubdomains
[!] Public Key Pinning (HPKP) - No
[+] Compression - No
[*] Session resumption - No
[*] Session tickets - Yes
[*] OCSP Stapling - No
[*] Next Protocol Negotiation (NPN) - Yes (http/1.1)
[*] SNI Required - Yes
[*] -----------------------------------------------------------------
[*] Report for server-52-84-237-113.sfo5.r.cloudfront.net (52.84.237.113)
[*] -----------------------------------------------------------------
[+] Overall rating: A+
[+] TLS 1.2 - Yes
[+] TLS 1.1 - Yes
[+] TLS 1.0 - Yes
[+] SSL 3.0 - No
[+] SSL 2.0 - No
[+] Secure renegotiation is supported
[!] BEAST attack - Yes
[+] POODLE SSLv3 - Not vulnerable
[+] POODLE TLS - Not vulnerable
[+] Downgrade attack prevention - Yes, TLS_FALLBACK_SCSV supported
[+] Freak - Not vulnerable
[+] RC4 - No
[*] Heartbeat (extension) - No
[+] Heartbleed (vulnerability) - No
[+] OpenSSL CCS vulnerability (CVE-2014-0224) - No
[+] Forward Secrecy - With modern browsers
[+] Strict Transport Security (HSTS) - Yes:max-age=15552000:includeSubdomains
[!] Public Key Pinning (HPKP) - No
[+] Compression - No
[*] Session resumption - No
[*] Session tickets - Yes
[*] OCSP Stapling - No
[*] Next Protocol Negotiation (NPN) - Yes (http/1.1)
[*] SNI Required - Yes
[*] -----------------------------------------------------------------
[*] Report for server-52-84-237-53.sfo5.r.cloudfront.net (52.84.237.53)
[*] -----------------------------------------------------------------
[+] Overall rating: A+
[+] TLS 1.2 - Yes
[+] TLS 1.1 - Yes
[+] TLS 1.0 - Yes
[+] SSL 3.0 - No
[+] SSL 2.0 - No
[+] Secure renegotiation is supported
[!] BEAST attack - Yes
[+] POODLE SSLv3 - Not vulnerable
[+] POODLE TLS - Not vulnerable
[+] Downgrade attack prevention - Yes, TLS_FALLBACK_SCSV supported
[+] Freak - Not vulnerable
[+] RC4 - No
[*] Heartbeat (extension) - No
[+] Heartbleed (vulnerability) - No
[+] OpenSSL CCS vulnerability (CVE-2014-0224) - No
[+] Forward Secrecy - With modern browsers
[+] Strict Transport Security (HSTS) - Yes:max-age=15552000:includeSubdomains
[!] Public Key Pinning (HPKP) - No
[+] Compression - No
[*] Session resumption - No
[*] Session tickets - Yes
[*] OCSP Stapling - No
[*] Next Protocol Negotiation (NPN) - Yes (http/1.1)
[*] SNI Required - Yes
[*] -----------------------------------------------------------------
[*] Report for server-52-84-237-48.sfo5.r.cloudfront.net (52.84.237.48)
[*] -----------------------------------------------------------------
[+] Overall rating: A+
[+] TLS 1.2 - Yes
[+] TLS 1.1 - Yes
[+] TLS 1.0 - Yes
[+] SSL 3.0 - No
[+] SSL 2.0 - No
[+] Secure renegotiation is supported
[!] BEAST attack - Yes
[+] POODLE SSLv3 - Not vulnerable
[+] POODLE TLS - Not vulnerable
[+] Downgrade attack prevention - Yes, TLS_FALLBACK_SCSV supported
[+] Freak - Not vulnerable
[+] RC4 - No
[*] Heartbeat (extension) - No
[+] Heartbleed (vulnerability) - No
[+] OpenSSL CCS vulnerability (CVE-2014-0224) - No
[+] Forward Secrecy - With modern browsers
[+] Strict Transport Security (HSTS) - Yes:max-age=15552000:includeSubdomains
[!] Public Key Pinning (HPKP) - No
[+] Compression - No
[*] Session resumption - No
[*] Session tickets - Yes
[*] OCSP Stapling - Yes
[*] Next Protocol Negotiation (NPN) - Yes (http/1.1)
[*] SNI Required - Yes
[*] -----------------------------------------------------------------
[*] Report for server-52-84-237-237.sfo5.r.cloudfront.net (52.84.237.237)
[*] -----------------------------------------------------------------
[+] Overall rating: A+
[+] TLS 1.2 - Yes
[+] TLS 1.1 - Yes
[+] TLS 1.0 - Yes
[+] SSL 3.0 - No
[+] SSL 2.0 - No
[+] Secure renegotiation is supported
[!] BEAST attack - Yes
[+] POODLE SSLv3 - Not vulnerable
[+] POODLE TLS - Not vulnerable
[+] Downgrade attack prevention - Yes, TLS_FALLBACK_SCSV supported
[+] Freak - Not vulnerable
[+] RC4 - No
[*] Heartbeat (extension) - No
[+] Heartbleed (vulnerability) - No
[+] OpenSSL CCS vulnerability (CVE-2014-0224) - No
[+] Forward Secrecy - With modern browsers
[+] Strict Transport Security (HSTS) - Yes:max-age=15552000:includeSubdomains
[!] Public Key Pinning (HPKP) - No
[+] Compression - No
[*] Session resumption - No
[*] Session tickets - Yes
[*] OCSP Stapling - Yes
[*] Next Protocol Negotiation (NPN) - Yes (http/1.1)
[*] SNI Required - Yes
[*] Auxiliary module execution completed
```

**Comment by h00die** [comment]
# Release Notes
This PR fixes up the ssl labs module to work with the new API and be more resilient to future changes

**Pull request merged**